punchcut/src/source/glyf/scaler.rs

```diff
@@ -0,0 +1,621 @@
+                    for p in &mut outline.points[point_end - 4..] {
+                        p.x = math::round(p.x);
+                        p.y = math::round(p.y);
+                    }
+                    // Apply hinting to the set of contours for this outline.
+                    if !self.hint(outline, point_base, contour_base, ins, false) {
+                        return Err(Error::HintingFailed(glyph_id));
+                    }
+                }
+                if point_base != 0 {
+                    // If we're not the first component, shift our contour end points.
+                    for c in &mut outline.contours[contour_base..contour_end] {
+                        *c += point_base as u16;
+                    }
+                }
+                // We're done with the phantom points, so drop them.
+                outline.points.truncate(outline.points.len() - 4);
+                outline.tags.truncate(outline.tags.len() - 4);
+            }
+            Glyph::Composite(composite) => {
+                if self.scaler.is_scaled {
+                    let scale = self.scaler.scale;
+                    for p in self.phantom.iter_mut() {
+                        p.x = math::mul(p.x, scale);
+                        p.y = math::mul(p.y, scale);
+                    }
+                }
+                // TODO: variations
+                // let delta_base = self.deltas.len();
+                // let mut have_deltas = false;
+                // let count = composite.components().count();
+                // self.deltas.resize(delta_base + count, Point::new(0, 0));
+                // if state.data.composite_deltas(
+                //     state.coords,
+                //     glyph_id,
+                //     &mut self.deltas[delta_base..],
+                // ) {
+                //     have_deltas = true;
+                // }
+                for component in composite.components() {
+                    // Save a copy of our phantom points.
+                    let phantom = self.phantom;
+                    // Load the component glyph and keep track of the points range.
+                    let start_point = outline.points.len();
+                    self.load(component.glyph, outline, recurse_depth + 1)?;
+                    let end_point = outline.points.len();
+                    if !component
+                        .flags
+                        .contains(CompositeGlyphFlags::USE_MY_METRICS)
+                    {
+                        // The USE_MY_METRICS flag indicates that this component's phantom
+                        // points should override those of the composite glyph.
+                        self.phantom = phantom;
+                    }
+                    // Scaling does an internal conversion to 26.6 so we don't use the
+                    // fixed types in read-fonts. Maybe a better solution here? We want to match
+                    // FreeType semantics.
```

Comment:
We can add some convenience methods to make this a bit nicer, but ultimately there's some inherent complexity around how both scaling and hinting handle a mix of integers and fixed 26.6 values. It might be worth just taking the time to identify and implement these things.